crates/red_knot_python_semantic/src/semantic_index/usedef.rs

```diff
@@ -0,0 +1,127 @@
+use crate::ast_node_ref::AstNodeRef;
+use crate::db::Db;
+use crate::semantic_index::ast_ids::ScopedUseId;
+use crate::semantic_index::definition::Definition;
+use crate::semantic_index::symbol::{FileScopeId, ScopeId, ScopedSymbolId};
+use ruff_db::files::File;
+use ruff_index::IndexVec;
+use ruff_python_ast as ast;
+use salsa;
+
+/// An independently type-inferable expression.
+///
+/// Includes constraint expressions (e.g. if tests) and the RHS of an unpacking assignment.
+#[salsa::tracked]
+pub(crate) struct Expression<'db> {
+    /// The file in which the expression occurs.
+    #[id]
+    pub(crate) file: File,
+
+    /// The scope in which the expression occurs.
+    #[id]
+    pub(crate) file_scope: FileScopeId,
+
+    /// The statement node containing the expression.
+    /// We hold the statement node to clarify which specific contexts these
+    /// independently-inferable expressions can occur in, and limit the variants of this enum.
+    #[no_eq]
+    #[return_ref]
+    pub(crate) ast: ExpressionKind,
+}
+
+impl<'db> Expression<'db> {
+    /// Return the actual expression.
+    #[allow(unused)]
+    pub(crate) fn node(self, db: &'db dyn Db) -> &ast::Expr {
+        match self.ast(db) {
+            ExpressionKind::AssignedValue(node_ref) => &node_ref.node().value,
+            ExpressionKind::IfTest(node_ref) => &node_ref.node().test,
```

Comment:
Should we just store the relevant node here, considering that we only need a part of it?

Reply:
I have a comment just above about this. In type inference we will need to recognize specific cases where we need to use the per-expression query, and so I thought it was clearer to explicitly store those specific cases, rather than generically store any expression.

Also, if I generically store any expression, then I'll need like a 15-variant enum here, for all the possible expression types. That seems more verbose and actually less clear about the "cases" we really deal with here.

I don't see a big downside to the current approach; following one box pointer doesn't seem like a big cost in the context of everything else?

Reply:
> I don't see a big downside to the current approach; following one box pointer doesn't seem like a big cost in the context of everything else?

No, it's not a big cost. It's just unclear to me how we avoid the 15 variants even if we hold on to the parent node, because you would still need different kinds to know which different sub-node to take. Anyway. I feel comfortable moving forward with this. It's just a design decision that I don't fully understand.

Reply:
Understood that this isn't a blocking issue, but following up anyway to see if we can clear up our common understanding of this design choice, in case I'm missing something:

> It's just unclear to me how we avoid the 15 variants even if we hold on to the parent node, because you would still need different kinds to know which different sub-node to take.

So the "15 variants" I'm talking about here is like one variant for `ExprName`, one for `ExprBinOp`, etc, for all the different types of expressions. Just because we don't have an `AstNodeRef` equivalent of `ExpressionRef`, as far as I know.

If we hold the parent, we don't need those 15 variants, we just need one variant for each possible type of parent node that requires an independently-inferable expression, which will be more like 3 or 4 variants: one for StmtAssign, and one for each kind of branch statement that can contain a test expression, so `If`, `While`... not sure if there are any others? For each of those variants, there's only one sub-expression we could possibly mean (the RHS `value` for StmtAssign, the `test` for the others). And we don't need to enumerate all possible expression types anywhere, since type inference already handles that match.

So this design is many fewer variants, and the variants we have are more semantically meaningful: rather than just being a bunch of boilerplate to handle all possible expression types, it actually enumerates each specific semantic context in which we need an `Expression` ingredient at all.

Let me know if that still isn't making sense, or where I'm missing something!

Reply:
An `AstNodeRef` should be able to store a reference to an `Expr`. In the end, `AstNodeRef` allows you to reference any reference that has the same lifetime as the parsed tree. There's no `where` clause restricting the type. The only constraint is that the thing stored on `AstNodeRef` is part of the AST.

Reply:
Oh! In that case I will look to simplify this; we may not need an enum here at all.